pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationService.java

```diff
@@ -83,13 +83,29 @@ public String authenticate(AuthenticationDataSource authData, String authMethodN
     }
 
     public String authenticateHttpRequest(HttpServletRequest request) throws AuthenticationException {
-        // Try to validate with any configured provider
         AuthenticationDataSource authData = new AuthenticationDataHttps(request);
-        for (AuthenticationProvider provider : providers.values()) {
+        String authMethodName = request.getHeader("Auth-Method-Name");
+
+        if (authMethodName != null) {
+            AuthenticationProvider providerToUse = providers
+                    .values()
+                    .parallelStream()
+                    .filter(provider -> provider.getAuthMethodName().equals(authMethodName))
+                    .findAny()
+                    .orElseThrow(() -> new AuthenticationException(String.format("Unsupported header Auth-Method-Name [%s].", authMethodName)));
+
             try {
-                return provider.authenticate(authData);
+                return providerToUse.authenticate(authData);
             } catch (AuthenticationException e) {
-                // Ignore the exception because we don't know which authentication method is expected here.
+                throw e;
+            }
+        } else {
+            for (AuthenticationProvider provider : providers.values()) {
+                try {
+                    return provider.authenticate(authData);
+                } catch (AuthenticationException e) {
+                    // Ignore the exception because we don't know which authentication method is expected here.
+                }
             }
```

Comment:
If none of the providers is able to authenticate the provided `authData`, shouldn't we throw any exception here?

Reply:
Already done here: https://github.com/apache/pulsar/blob/562231c7faf88fd1fed1ce4ac2dc7b5084a93896/pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationService.java#L118, right?